http/server.go

```diff
@@ -0,0 +1,141 @@
+package http
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-log/v2"
+
+	"github.com/drand/drand/beacon"
+	"github.com/drand/drand/key"
+	"github.com/drand/drand/protobuf/drand"
+)
+
+var (
+	// Timeout for how long to wait for the drand.PublicClient before timing out
+	reqTimeout = 5 * time.Second
+)
+
+// New creates an HTTP handler for the public Drand API
+func New(ctx context.Context, client drand.PublicClient) (http.Handler, error) {
+	pkt, err := client.Group(ctx, &drand.GroupRequest{})
+	if err != nil {
+		return nil, err
+	}
+	if pkt == nil {
+		return nil, fmt.Errorf("Failed to retrieve valid GroupPacket")
+	}
+	parsedPkt, err := key.GroupFromProto(pkt)
+	if err != nil {
+		return nil, err
+	}
+
+	handler := handler{reqTimeout, client, parsedPkt, log.Logger("http")}
+
+	mux := http.NewServeMux()
+	//TODO: aggregated bulk round responses.
+	mux.HandleFunc("/public/latest", handler.LatestRand)
+	mux.HandleFunc("/public/", handler.PublicRand)
+	mux.HandleFunc("/group", handler.Group)
+	return mux, nil
+}
+
+type handler struct {
+	timeout   time.Duration
+	client    drand.PublicClient
+	groupInfo *key.Group
+	log       log.StandardLogger
+}
+
+func (h *handler) getRand(round uint64) ([]byte, error) {
+	req := drand.PublicRandRequest{Round: round}
+	ctx, cancel := context.WithTimeout(context.Background(), h.timeout)
+	defer cancel()
+	resp, err := h.client.PublicRand(ctx, &req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(resp)
+}
+
+func (h *handler) PublicRand(w http.ResponseWriter, r *http.Request) {
+	// Get the round.
+	round := strings.Replace(r.URL.Path, "/public/", "", 1)
+	roundN, err := strconv.ParseUint(round, 10, 64)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		h.log.Warnf("%s %d - %s", r.RemoteAddr, http.StatusBadRequest, url.PathEscape(r.URL.Path))
+		return
+	}
+
+	data, err := h.getRand(roundN)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		h.log.Warnf("%s %d - %s %v", r.RemoteAddr, http.StatusInternalServerError, url.PathEscape(r.URL.Path), err)
+		return
+	}
+
+	roundExpectedTime := beacon.TimeOfRound(h.groupInfo.Period, h.groupInfo.GenesisTime, roundN)
+
+	http.ServeContent(w, r, "rand.json", time.Unix(roundExpectedTime, 0), bytes.NewReader(data))
+}
+
+func (h *handler) LatestRand(w http.ResponseWriter, r *http.Request) {
+	req := drand.PublicRandRequest{Round: 0}
+	ctx, cancel := context.WithTimeout(context.Background(), h.timeout)
+	defer cancel()
+
+	resp, err := h.client.PublicRand(ctx, &req)
```

Comment:
Question: We could be smarter about this latest: compute the current time and look the last beacon we had and if the current round is higher than that, actually making the connection ?
Am i understanding right this code that each time I call `latest` a connection is actually made on the drand node ?

Reply:
I want the CDN in front of this code to be smart and do the caching. Currently, every call to this code will make a connection. However, the point is it also tells the caller explicitly the next time it's worth reconnecting to ask for a new one.

As long as the CDN respects that, i'd rather not have the extra cache at this level, because excess caching can lead to overly-slow propagation of values.